### PR TITLE
Unconditionally show update nightly hint on ICE

### PR DESCRIPTION
**Pull request by jieyouxu**
Instead of trying to guess if a update nightly hint should be shown (by checking for system time, querying version and channel info etc.), just show the update nightly hint for nightly compilers. This avoids breaking tests that match on ICE test outputs on nightly/dev channels.

> Another issue is that the outdated nightly hint triggers for ICE tests, causing a mismatch with the test expectation. There doesn't seem to be any env var to suppress this.

See <https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/stage0.20compiletest.20broken/near/425543681> for context.

### CODE REVIEW

**Comment by onur-ozkan** [comment]
We should link this PR with #122196 if I am not mistaken?

**Comment by jieyouxu** [comment]
> We should link this PR with #122196 if I am not mistaken?

Unfortunately there's 2 separate bugs here:
1. The ICE message is shown conditionally on a lot of conditions which trigger for some ICE tests when tested via outdated nightly compilers.
2. #122196 is still broken because `rmake.rs` isn't compiled with `--sysroot` passed.

**Comment by cjgillot** [comment]
I'm lacking context about this hint.
r? compiler

**Comment by lcnr** [comment]
same :sweat_smile: 
r? compiler

File: tests/ui/impl-trait/equality-in-canonical-query.clone.stderr [hunk 1]
```diff
@@ -19,5 +19,7 @@ LL |     same_output(foo, rpit);
 LL |     same_output(foo, rpit);
    |     ^^^^^^^^^^^^^^^^^^^^^^
 
+note: if you are using a nightly compiler, please make sure that you have updated to the latest nightly
```
Comment (estebank) [inline tests/ui/impl-trait/equality-in-canonical-query.clone.stderr:22]: Is there any point in having "if you are using a nightly compiler"? Why not check `tcx.sess.is_nightly_build()` directly?

**Comment by estebank** [comment]
I feel like we would want to unconditionally suggest updating for nightly releases. On stable I am of many minds. We could do a check for current date and try to figure out how long ago the current release went out, and if >6 weeks, tell the user to consider updating. But that wouldn't take into account dot-releases to fix that exact ICE being released. At the same time, I would expect people to already think "maybe there's a newer version that fixes this". On stable we also need to account for non-rustup distributed rustc, like from Linux distro package managers, which would likely not want such a message, or to customize it...

For now, can you try to check for the current build release channel and remove the uncertain language from the message?

File: tests/ui/impl-trait/equality-in-canonical-query.clone.stderr (hunk 1 above)
Comment (jieyouxu) [reply tests/ui/impl-trait/equality-in-canonical-query.clone.stderr:22]: I thought install_ice_hook and report_ice were so early in rustc_driver_impl that we didn't have a tcx available?

**Comment by jieyouxu** [comment]
> For now, can you try to check for the current build release channel and remove the uncertain language from the message?

Hmm, wouldn't this break ICE tests checking for the exact ICE message if the compiler's channel is `nightly`? (because normally the ICE tests are blessed under some `dev` channel right?)

**Comment by jieyouxu** [comment]
@estebank sorry for the ping, I want to unbork cg_clif's tests. Not sure what to do here, maybe we just emit the update nightly note unconditionally, or revert the update note altogether?

File: tests/ui/impl-trait/equality-in-canonical-query.clone.stderr (hunk 1 above)
Comment (estebank) [reply tests/ui/impl-trait/equality-in-canonical-query.clone.stderr:22]: I believe you can use `rustc_feature::UnstableFeatures::from_environment(None).is_nightly_build();` for this instead.

**Comment by estebank** [comment]
@jieyouxu let's check if `rustc_feature::UnstableFeatures::from_environment(None).is_nightly_build()` works for checking that the ICE isn't happening on a stable release. If it doesn't work, lets land this PR as is.

**Comment by jieyouxu** [comment]
Thanks, I'll try that

**Pull request merged**